### ISSUE TITLE
ANR in Firebase Performance 19.09

### ISSUE THREAD

**Issue by benkay** [firebase/firebase-android-sdk#2066]
### [READ] Step 1: Are you in the right place?
✅ 

### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: 4.1.0
  * Firebase Component: Performance (Database, Firestore, Storage, Functions, etc)
  * Component version: 19.0.9

### [REQUIRED] Step 3: Describe the problem

We tried to update from firebase perf 19.0.8 -> 19.0.9, but we're seeing a pretty regular deadlock on app start.

#### Steps to reproduce:

What happened? How can we make the problem occur?
This could be a description, log/console output, etc.

Happens fairly frequently on a normal cold start - on some devices it's almost every start. Relevant ANR traces are below. 

```
"main" prio=5 tid=1 Blocked
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x72d62448 self=0xb4000074dad347b0
  | sysTid=26830 nice=-10 cgrp=default sched=0/0 handle=0x76015b24f8
  | state=S schedstat=( 75018076 10181923 288 ) utm=4 stm=2 core=6 HZ=100
  | stack=0x7fd9dcf000-0x7fd9dd1000 stackSize=8192KB
  | held mutexes=
  at com.google.firebase.perf.internal.FirebasePerfClearcutLogger.getInstance(SourceFile:2)
  - waiting to lock <0x0f5071bc> (a java.lang.Class<m.h.c.v.g.d>) held by thread 55
  at com.google.firebase.perf.FirebasePerformance.<init>(SourceFile:2)
  at com.google.firebase.perf.FirebasePerfRegistrar.lambda$getComponents$0(SourceFile:4)
  at com.google.firebase.perf.FirebasePerfRegistrar$$Lambda$1.create(SourceFile:-1)
  at com.google.firebase.components.ComponentRuntime.lambda$new$0(SourceFile:2)
  at com.google.firebase.components.ComponentRuntime$$Lambda$1.get(SourceFile:-1)
  at com.google.firebase.components.Lazy.get(SourceFile:6)
  - locked <0x0b4fe545> (a m.h.c.j.t)
  at com.google.firebase.FirebaseApp.com.google.firebase.components.ComponentRuntime.initializeEagerComponents(SourceFile:18)
                                     initializeAllApis
  at com.google.firebase.FirebaseApp.initializeApp(SourceFile:13)
  at com.google.firebase.provider.FirebaseInitProvider.com.google.firebase.FirebaseApp.initializeApp(SourceFile:7)
                                                       com.google.firebase.FirebaseApp.initializeApp
                                                       onCreate
  - locked <0x013dd49a> (a java.lang.Object)
  at android.content.ContentProvider.attachInfo(ContentProvider.java:2388)
  at android.content.ContentProvider.attachInfo(ContentProvider.java:2358)
  at com.google.firebase.provider.FirebaseInitProvider.attachInfo(SourceFile:3)
  at android.app.ActivityThread.installProvider(ActivityThread.java:7239)
  at android.app.ActivityThread.installContentProviders(ActivityThread.java:6780)
  at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6697)
  at android.app.ActivityThread.access$1300(ActivityThread.java:237)
  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1913)
  at android.os.Handler.dispatchMessage(Handler.java:106)
  at android.os.Looper.loop(Looper.java:223)
  at android.app.ActivityThread.main(ActivityThread.java:7656)
  at java.lang.reflect.Method.invoke(Native method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)

"HttpRequestManager:4178" prio=1 tid=55 Blocked
  | group="HttpRequestManager" sCount=1 dsCount=0 flags=1 obj=0x12c43600 self=0x74dad76890
  | sysTid=26936 nice=19 cgrp=default sched=0/0 handle=0x72ecab6cc0
  | state=S schedstat=( 1201042 32969 5 ) utm=0 stm=0 core=5 HZ=100
  | stack=0x72ec9b3000-0x72ec9b5000 stackSize=1043KB
  | held mutexes=
  at com.google.firebase.FirebaseApp.getInstance(SourceFile:1)
  - waiting to lock <0x013dd49a> (a java.lang.Object) held by thread 1
  at com.google.firebase.perf.internal.FirebasePerfClearcutLogger.getInstance(SourceFile:4)
  - locked <0x0f5071bc> (a java.lang.Class<m.h.c.v.g.d>)
  at com.google.firebase.perf.network.FirebasePerfUrlConnection.instrument(SourceFile:3)
  at com.flurry.sdk.df.d(SourceFile:9)
  at com.flurry.sdk.df.a(SourceFile:7)
  at com.flurry.sdk.dd.a(SourceFile:3)
  at com.flurry.sdk.eb.run(SourceFile:1)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
  at java.lang.Thread.run(Thread.java:923)
```

#### Relevant Code:

```
n/a
```

**Comment by alikn** [comment]
Hi @benkay, 
Thanks for reporting this issue. I tried reproducing it on a blank app but could not. Was wondering if you mind sharing some info in order to help me reproduce?
Here are some of the things which could be helpful. Feel free to skip any of them which you cannot share.

1. Other dependencies of the app (maybe the gradle file)
2. A sample app on which the error could be reproduced.

**Comment by benkay** [comment]
Sorry, didn't get time to respond in the last few days.
I could try to put together a sample app, but it looks like a race condition when Flurry's content provider fires off a network request on a background thread, and meanwhile FirebaseInitProvider tries to initialize itself.

Not actually sure why we're only seeing it in 19.0.9 as it looks like the underlying deadlock is there in previous versions too - perhaps something's caused the initialization order to change?

1. FirebaseApp.initializeApp() is called, and takes a lock internally 
2. Flurry's network call is intercepted by firebase perf which calls FirebasePerfClearcutLogger.getInstance(). This is not yet initialized so synchronizes its own class, then tries to call FirebaseApp.getInstance() which ends up waiting on the lock that the initializeApp() thread holds
3. At some point in the Firebase initialization FirebasePerfClearcutLogger.getInstance() is called, and gets stuck waiting for the network call thread 💥 

Not sure why FirebasePerfClearcutLogger.getInstance() tries to call FirebaseApp.getInstance() (seemingly only for the purposes of bailing out if it's not yet initialized?), but moving that before the synchronized block should avoid the deadlock?

**Comment by jposuna** [comment]
Hi @benkay we have a change that we believe will fix this issue which will be shipped in the next release. Once the new version is available, please try it out and let us know if it doesn't fix your issue.

I will close the ticket in the meantime, but please feel free to reopen it if the new version does not fix the issue.

**Comment by riggaroo** [comment]
Which release will this be fixed in? I see there is a 19.0.10 but it doesn't say it includes this change.
